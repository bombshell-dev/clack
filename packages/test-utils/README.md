# `@clack/test-utils`

A bunch of useful test utiltiies for use inside clack.
