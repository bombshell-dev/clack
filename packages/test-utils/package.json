{
  "name": "@clack/test-utils",
  "private": true,
  "version": "1.0.0-alpha.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bombshell-dev/clack.git",
    "directory": "packages/test-utils"
  },
  "bugs": {
    "url": "https://github.com/bombshell-dev/clack/issues"
  },
  "homepage": "https://github.com/bombshell-dev/clack/tree/main/packages/test-utils#readme",
  "files": ["dist", "CHANGELOG.md"],
  "author": {
    "name": "James Garbutt",
    "url": "https://github.com/43081j"
  },
  "license": "MIT",
  "packageManager": "pnpm@9.14.2",
  "scripts": {
    "build": "tsc",
    "prepack": "pnpm build"
  },
  "dependencies": {
  },
  "devDependencies": {
  }
}
