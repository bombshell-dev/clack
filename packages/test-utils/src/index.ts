export { MockReadable } from './mock-readable.js';
export { MockWritable } from './mock-writable.js';

export function nextTick(): Promise<void> {
	return new Promise<void>((resolve) => {
		setTimeout(resolve, 0);
	});
}
