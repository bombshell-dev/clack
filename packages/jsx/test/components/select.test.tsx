import { MockReadable, MockWritable } from '@clack/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';
import { Option, Select } from '../../src/index.js';

describe('Select', () => {
	let input: MockReadable;
	let output: MockWritable;

	beforeEach(() => {
		input = new MockReadable();
		output = new MockWritable();
	});

	test('renders options', async () => {
		const task = (
			<Select message="foo" input={input} output={output}>
				<Option value={'opt0'} />
				<Option value={'opt1'} />
			</Select>
		);

		// wait a tick... sad times
		await new Promise((res) => setTimeout(res, 0));

		input.emit('keypress', '', { name: 'return' });

		const result = await task;

		expect(result).to.equal('opt0');
		expect(output.buffer).toMatchSnapshot();
	});

	test('renders options with labels', async () => {
		const task = (
			<Select message="foo" input={input} output={output}>
				<Option value={303}>Three o three</Option>
				<Option value={808}>Eight o eight</Option>
			</Select>
		);

		// wait a tick... sad times
		await new Promise((res) => setTimeout(res, 0));

		input.emit('keypress', '', { name: 'return' });

		const result = await task;

		expect(result).to.equal(303);
		expect(output.buffer).toMatchSnapshot();
	});

	test('renders options with hints', async () => {
		const task = (
			<Select message="foo" input={input} output={output}>
				<Option value={303} hint="hint one">
					Three o three
				</Option>
				<Option value={808} hint="hint two">
					Eight o eight
				</Option>
			</Select>
		);

		// wait a tick... sad times
		await new Promise((res) => setTimeout(res, 0));

		input.emit('keypress', '', { name: 'return' });

		const result = await task;

		expect(result).to.equal(303);
		expect(output.buffer).toMatchSnapshot();
	});
});
