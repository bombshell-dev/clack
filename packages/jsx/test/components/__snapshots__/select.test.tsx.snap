// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > renders options 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m opt0
[36m│[39m  [2m○[22m [2mopt1[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=5>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mopt0[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`Select > renders options with hints 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Three o three [2m(hint one)[22m
[36m│[39m  [2m○[22m [2mEight o eight[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=5>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mThree o three[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`Select > renders options with labels 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Three o three
[36m│[39m  [2m○[22m [2mEight o eight[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=5>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mThree o three[22m",
  "
",
  "<cursor.show>",
]
`;
